### PR TITLE
Create mesh

### PR DESCRIPTION
**Pull request by samwaseda**
This module allows us to create mesh grid from atomistic structures that can be used in continuum calculations

## Summary by CodeRabbit

- **New Features**
  - Added `create_mesh` functionality to generate a mesh grid from atomic structures.

- **Tests**
  - Added unit tests for the `create_mesh` function to ensure reliable performance with various parameters.

### CODE REVIEW

**Comment by liamhuber** [comment]
```
n_mesh: int|list = 10,
      TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

`|` for type hinting was not introduced until 3.10 IIRC. `from __future__ import annotations` helps with a variety of type hint issues, but I don't remember if it helps with this one. You might need to deprecate a python version or switch to the uglier `typing.Union[]` and `typing.Optional[]`

File: structuretoolkit/build/mesh.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    Returns:
+        (n, n, n, 3)-array: mesh
+    """
+    if n_mesh is None:
+        if density is None:
+            raise ValueError("either n_mesh or density must be specified")
+        n_mesh = np.rint(np.linalg.norm(structure.cell, axis=-1) / density).astype(int)
+    elif density is not None:
+        warnings.warn("As n_mesh is not `None`, `density` is ignored")
```
Comment (liamhuber) [inline structuretoolkit/build/mesh.py:33]: Why not fail hard? If they've bothered to specify a density, they've either forgotten to turn off the `n_mesh` and probably don't want this to run and would appreciate it not to, or they've intentionally specified both because they misunderstand the basics and we should slap them in the face with an error anyhow.

File: structuretoolkit/build/mesh.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        n_mesh (int): Number of grid points in each direction. If one number
+            is given, it will be repeated in every direction (i.e. n_mesh = 3
+            is the same as n_mesh = [3, 3, 3])
+        density (float): Density of grid points. Ignored when n_mesh is not
+            None
+        endpoint (bool): Whether both the edges get separate points or not.
+            cf. endpoint in numpy.linspace
+
+    Returns:
+        (n, n, n, 3)-array: mesh
```
Comment (liamhuber) [inline structuretoolkit/build/mesh.py:26]: Does it though? What happens when I invoke `create_mesh(some_structure, n_mesh=(4, 4, 4, 4))`? I'm not sure how the einsum will handle it, but up until that point the rest of the code looks like it will happily give you a hyper mesh. Same problem with `n_mesh=(4,4)`, although there I'm more confident the einsum will fail because it over-specifies indices (although it's not obvious to me that the error message will be clear and helpful).

If you want this to be restricted to three-dimensional data, some checking of `n_mesh` will be required.

File: structuretoolkit/build/mesh.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    structure: Atoms,
+    n_mesh: int|list = 10,
+    density: float|None = None,
+    endpoint: bool = False
+):
+    """
+    Create a mesh based on a structure
+
+    Args:
+        structure (ase.atoms.Atoms): ASE Atoms
```
Comment (liamhuber) [inline structuretoolkit/build/mesh.py:16]: Would-be-nice: to overload this argument to something like (np.ndarray | ase.atoms.Atoms), and maybe to re-name it `cell`. This function is in `structuretoolkit`, so I won't begrudge you calling it `structure` and making that the priority input, but this is actually a super generic function and should work on any 3x3 cell (maybe with some verification that the cell vectors follow some right hand rule or something, I haven't thought too deeply). 

Anyhow, a sort of multiple-dispatch approach for overloading the input types might be friendly to users.

**Comment by samwaseda** [comment]
> I'm grumpy being asked for review when the unit tests fail 😝 but overall looks nice, I basically just have some complaints about gracefully handling edge cases.

Yeah XD I didn't expect it to fail because I had run the tests.

**Comment by samwaseda** [comment]
Is it actually possible that the definition of `(nx, ny, nz, 3)` is not compatible with the [`RectMesh` definition](https://github.com/pyiron/pyiron_continuum/blob/main/pyiron_continuum/mesh.py)? I got the impression that there, it's `(3, nx, ny, nz)`?

**Comment by liamhuber** [comment]
> Is it actually possible that the definition of `(nx, ny, nz, 3)` is not compatible with the [`RectMesh` definition](https://github.com/pyiron/pyiron_continuum/blob/main/pyiron_continuum/mesh.py)? I got the impression that there, it's `(3, nx, ny, nz)`?

Yep, looks like:

```python
from pyiron_continuum.mesh import RectMesh
m = RectMesh(bounds=(1,1,1), divisions=(4,5,6))
m.shape
>>> (3, 4, 5, 6)
```

[Sounds like I did that](https://github.com/pyiron/pyiron_continuum/blob/9a633ada41ad719324cd6f54533af93c4e9de6de/pyiron_continuum/mesh.py#L63) so that the real-space dimension always appears first. Otherwise it sounds like 2d is some weird special case?

Consistency is definitely something to shoot for, but these functions actually appear in totally different packages, so if their behaviour differs it is not a deal breaker for me.

**Comment by samwaseda** [comment]
> Consistency is definitely something to shoot for, but these functions actually appear in totally different packages, so if their behaviour differs it is not a deal breaker for me.

Well since I’m doing it for multiscale modeling, it’s a total deal breaker if the definition is not the same. I’m gonna change it later.

File: structuretoolkit/build/mesh.py (hunk 3 above)
Comment (samwaseda) [reply structuretoolkit/build/mesh.py:17]: That's actually a good idea. I'm gonna go for it.

File: structuretoolkit/build/mesh.py (hunk 1 above)
Comment (samwaseda) [reply structuretoolkit/build/mesh.py]: I changed it

File: structuretoolkit/build/mesh.py (hunk 2 above)
Comment (samwaseda) [reply structuretoolkit/build/mesh.py]: Now it checks the shape

File: structuretoolkit/build/mesh.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            None
+        endpoint (bool): Whether both the edges get separate points or not.
+            cf. endpoint in numpy.linspace
+
+    Returns:
+        (n, n, n, 3)-array: mesh
+    """
+    if n_mesh is None:
+        if density is None:
+            raise ValueError("either n_mesh or density must be specified")
```
[the remark this answers is not in the document]
Comment (samwaseda) [reply structuretoolkit/build/mesh.py]: Done

**Review by liamhuber — APPROVED** [review]
Nice!
